### DISCUSSION TITLE
Support derivation to non-sealed trait

### DISCUSSION THREAD

**Discussion by jtjeferreira** [scalalandio/chimney#561] [General]
Hi!

Thanks for this amazing library.

Consider the following code

```scala
import io.scalaland.chimney.Transformer
import io.scalaland.chimney.inlined._

sealed trait Foo
case object A extends Foo

sealed trait Bar
case object B extends Bar

trait Baz
case object C extends Baz


val barT = Transformer.define[Foo, Bar].withSealedSubtypeRenamed[A.type, B.type].buildTransformer //works
val bazT = Transformer.define[Foo, Baz].withSealedSubtypeRenamed[A.type, C.type].buildTransformer //fails

```

The definition of a `Transformer` from `Foo` to `Bar` works. But the definition from `Foo` to `Baz` does not because `Baz` is not a sealed trait. 

I understand that because `Baz` is not sealed at compile time we can't find `C`, so there is not way to get `Transformer.define[Foo, Baz].buildTransformer` to work. But it should be possible to make `Transformer.define[Foo, Baz].withSealedSubtypeRenamed[A.type, C.type].buildTransformer` to work, right?

**Comment by MateuszKubuszok** [comment]
It could, but I would be worried to enable such thing without a lot of tests to make sure it would not have some unpredictable consequences:

 * [SealedHierarchyToSealedHierarchyRule](https://github.com/scalalandio/chimney/blob/master/chimney/src/main/scala/io/scalaland/chimney/internal/compiletime/derivation/transformer/rules/TransformSealedHierarchyToSealedHierarchyRuleModule.scala#L16) would have to detect case when ttarget type is `trait`/`abstract class` without sealed
   * there is not cross platform utility for checking that yet, so it would have to be developed
 * synthesize empty map of known types
 * test what would break

**Reply by jtjeferreira** [reply]
Thanks for the reply.
I will give this a try in the next days and submit a PR with the results.

**Reply by jtjeferreira** [reply]
Hi @MateuszKubuszok, I gave this a try in https://github.com/scalalandio/chimney/pull/564.

> SealedHierarchyToSealedHierarchyRule would have to detect case when ttarget type is `trait`/`abstract class` without sealed

I didn't implement this restriction, because I think it should be possible to configure a transformation where the target is a class (or any other type for that matter; in the example in the PR I used a `String`). But lets continue the discussion in the PR

**Reply by MateuszKubuszok** [reply]
Hi @jtjeferreira. Thanks! I'll review the PR later on today or tomorrow.

**Comment by jtjeferreira** [comment]
As per the comments in https://github.com/scalalandio/chimney/pull/564#discussion_r1661380862 this can be achieved without using chimney in a simpler way.
